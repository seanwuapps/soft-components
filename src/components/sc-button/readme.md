# sc-button


<!-- Auto Generated Below -->


## Usage

### Sc-button

<div class="intro">
A button is the most identifiable component for neumorphism designs. You can customise the use of a button in various ways. 
</div>

```html
<sc-button aria-label="Default button" onclick="alert('yo')">Button</sc-button>
```

You can use it the same way as an `<a>` tag.

```html
<!-- Anchor -->
<sc-button aria-label="Anchor tag" href="/" target="_blank" rel="noreferrer" title="button">
Anchor
</sc-button>
```

Add a `block` attribute to create a block level button.

```html
<!-- Block level -->
<sc-button aria-label="Block level button" block>Block Button</sc-button>
```

Use `icon` attribute to adjust styles to fit an icon into the button.

```html
<sc-button aria-label="Icon button" icon>
  <i class='bx bx-drink'></i>
</sc-button>
```

Use `icon-text` attribute to center align icon and text.

```html
<sc-button aria-label="Icon and text button" icon-text>
  <i class='bx bx-drink'></i>
  Text
</sc-button>

<sc-button aria-label="Icon and text button" icon-text>
  Text
  <i class='bx bx-drink'></i>
</sc-button>
```
Use `bordered` attribute to add a reflective border

```html
<sc-button aria-label="Icon button with a reflective border" bordered icon>
  <i class='bx bx-drink'></i>
</sc-button>
```

Use `circle` attribute to create a circle shaped button

```html
<sc-button aria-label="Circled button with image" circle>
  <img src="https://github.com/seanwuapps.png" alt="avatar" width="100%">
</sc-button>
```

Use `flat` attribute to make button blend in with the surface (not a11y friendly)

```html
<sc-button aria-label="Flat button" flat>
  Flat button
</sc-button>
```

Use `disabled` attribute to disable the button like you're used to.

```html
<sc-button aria-label="Disabled button" disabled>
  Disabled button
</sc-button>
```


Combine attributes together to have a party

```html

<h6>Circle icon</h6>
<sc-button aria-label="Circle icon" icon circle>
  <i class='bx bx-drink'></i>
</sc-button>

<h6>Circle bordered</h6>
<sc-button aria-label="Circle bordered" circle bordered>
  <img src="https://github.com/seanwuapps.png" alt="avatar" width="100%">
</sc-button>
```

You can add `ray-tracing` attribute to make it use the mouse as the light source, this could be used to draw more attention to the element.
```html
<sc-button ray-tracing>
  Look at me!
</sc-button>
```



## Properties

| Property     | Attribute     | Description                                                                                                                                                                                                                                                                               | Type                              | Default     |
| ------------ | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------- | ----------- |
| `active`     | `active`      | Set active state for the button                                                                                                                                                                                                                                                           | `boolean`                         | `false`     |
| `block`      | `block`       | Make button `display: block`                                                                                                                                                                                                                                                              | `boolean`                         | `false`     |
| `bordered`   | `bordered`    | If prop exists, button will have an engraved-styled border                                                                                                                                                                                                                                | `boolean`                         | `false`     |
| `circle`     | `circle`      | Make button circle shaped                                                                                                                                                                                                                                                                 | `boolean`                         | `false`     |
| `disabled`   | `disabled`    | If `true`, the user cannot interact with the button.                                                                                                                                                                                                                                      | `boolean`                         | `false`     |
| `download`   | `download`    | This attribute instructs browsers to download a URL instead of navigating to it, so the user will be prompted to save it as a local file. If the attribute has a value, it is used as the pre-filled file name in the Save prompt (the user can still change the file name if they want). | `string`                          | `undefined` |
| `flat`       | `flat`        | Make button flat                                                                                                                                                                                                                                                                          | `boolean`                         | `false`     |
| `href`       | `href`        | Contains a URL or a URL fragment that the hyperlink points to. If this property is set, an anchor tag will be rendered.                                                                                                                                                                   | `string`                          | `undefined` |
| `icon`       | `icon`        | Icon only button                                                                                                                                                                                                                                                                          | `boolean`                         | `false`     |
| `iconText`   | `icon-text`   | If this button has both icon and text                                                                                                                                                                                                                                                     | `boolean`                         | `false`     |
| `rayTracing` | `ray-tracing` | Use mouse as the light source (ray-tracing)                                                                                                                                                                                                                                               | `boolean`                         | `false`     |
| `rel`        | `rel`         | Specifies the relationship of the target object to the link object. The value is a space-separated list of [link types](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types).                                                                                                    | `string`                          | `undefined` |
| `target`     | `target`      | Specifies where to display the linked URL. Only applies when an `href` is provided. Special keywords: `"_blank"`, `"_self"`, `"_parent"`, `"_top"`.                                                                                                                                       | `string`                          | `undefined` |
| `type`       | `type`        | The type of the button.                                                                                                                                                                                                                                                                   | `"button" \| "reset" \| "submit"` | `'button'`  |


## Events

| Event        | Description                          | Type                |
| ------------ | ------------------------------------ | ------------------- |
| `blurEvent`  | Emitted when the button loses focus. | `CustomEvent<void>` |
| `clickEvent` | Emitted when the button is clicked.  | `CustomEvent<void>` |
| `focusEvent` | Emitted when the button has focus.   | `CustomEvent<void>` |


## Slots

| Slot | Description                                                           |
| ---- | --------------------------------------------------------------------- |
|      | Content is placed between the named slots if provided without a slot. |


## CSS Custom Properties

| Name                                 | Description                                                            |
| ------------------------------------ | ---------------------------------------------------------------------- |
| `--sc-button-bg-color`               | Background of the button. - default: var(--sc-bg-color, #f6f6f8)       |
| `--sc-button-border-radius`          | Button border radius. - default: var(--sc-border-radius, 1em)          |
| `--sc-button-bordered-border-radius` | Bordered-style border radius - default: var(--sc-button-border-radius) |
| `--sc-button-bordered-border-width`  | Bordered-style border width - default: 2px                             |
| `--sc-button-circle-size`            | Circle button size. - default: calc(var(--sc-root-spacing, 8px) * 6)   |
| `--sc-button-edge-width`             | Width of the edge - default: 4px                                       |
| `--sc-button-icon-size`              | Icon button size. - default: calc(var(--sc-root-spacing, 8px) * 6)     |
| `--sc-button-padding-x`              | Horizontal paddings. - default: calc(var(--sc-root-spacing, 8px) * 4)  |
| `--sc-button-padding-y`              | Vertical paddings. - default: calc(var(--sc-root-spacing, 8px) * 2)    |
| `--sc-button-text-active-color`      | Text color with active prop - default: var(--sc-active-color, #c586c0) |
| `--sc-button-text-color`             | Button text color. - default: var(--sc-text-color, #333333)            |
| `--sc-button-text-hover-color`       | Text color on hover - default: var(--sc-secondary-color, #646695)      |
| `--sc-button-vertical-align`         | vertical align of the button - default: top                            |
| `--sc-button-width`                  | Set fixed width for button - default: auto                             |


## Dependencies

### Depends on

- [sc-ray-tracer](../sc-ray-tracer)

### Graph
```mermaid
graph TD;
  sc-button --> sc-ray-tracer
  style sc-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
