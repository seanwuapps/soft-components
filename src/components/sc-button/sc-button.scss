sc-button {
  /**
    @prop --sc-button-bg-color:
      Background of the button.
      - default: var(--sc-bg-color, #f6f6f8)
    @prop --sc-button-text-color:
      Button text color.
      - default: var(--sc-text-color, #333333)
    @prop --sc-button-border-radius:
      Button border radius.
      - default: var(--sc-border-radius, 1em)
    @prop --sc-button-padding-y:
      Vertical paddings.
      - default: calc(var(--sc-root-spacing, 8px) * 2)
    @prop --sc-button-padding-x:
      Horizontal paddings.
      - default: calc(var(--sc-root-spacing, 8px) * 4)
    @prop --sc-button-icon-size:
      Icon button size.
      - default: calc(var(--sc-root-spacing, 8px) * 6)
    @prop --sc-button-circle-size:
      Circle button size.
      - default: calc(var(--sc-root-spacing, 8px) * 6)
    @prop --sc-button-bordered-border-radius:
      Bordered-style border radius
      - default: var(--sc-button-border-radius)
    @prop --sc-button-bordered-border-width:
      Bordered-style border width
      - default: 2px
    @prop --sc-button-text-hover-color:
      Text color on hover
      - default: var(--sc-secondary-color, #646695)
    @prop --sc-button-text-active-color:
      Text color with active prop
      - default: var(--sc-active-color, #c586c0)
    @prop --sc-button-width:
      Set fixed width for button
      - default: auto
    @prop --sc-button-edge-width:
      Width of the edge
      - default: 4px
    @prop --sc-button-vertical-align:
      vertical align of the button
      - default: top
    */

  --sc-button-bg-color: var(--sc-bg-color, #f6f6f8);
  --sc-button-text-color: var(--sc-text-color, #333333);
  --sc-button-border-radius: var(--sc-border-radius, 1em);
  --sc-button-padding-y: calc(var(--sc-root-spacing, 8px) * 2);
  --sc-button-padding-x: calc(var(--sc-root-spacing, 8px) * 4);
  --sc-button-icon-size: calc(var(--sc-root-spacing, 8px) * 6);
  --sc-button-circle-size: calc(var(--sc-root-spacing, 8px) * 6);
  --sc-button-bordered-border-radius: var(--sc-button-border-radius);
  --sc-button-bordered-border-width: 2px;
  --sc-button-text-hover-color: var(--sc-secondary-color, #646695);
  --sc-button-text-active-color: var(--sc-active-color, #c586c0);
  --sc-button-width: auto;
  --sc-button-edge-width: 4px;
  --sc-button-vertical-align: top;
  //-------------sc-button---------------//
  display: inline-block;
  vertical-align: var(--sc-button-vertical-align);
  &[type] {
    -webkit-appearance: none;
  }
  * {
    box-sizing: border-box;
  }

  &:not(.flat) {
    a,
    button {
      @include raised();
    }
  }
  a,
  button {
    cursor: pointer;
    border: none;
    text-decoration: none;
    display: block;
    background-color: var(--sc-button-bg-color);
    color: var(--sc-button-text-color);
    border-radius: var(--sc-button-border-radius);
    position: relative;
    padding: var(--sc-button-edge-width);
    font-size: 1em;
    font-family: inherit;
    width: var(--sc-button-width);

    .button-inner {
      overflow: hidden;
      transition: all 0.4s ease;
      display: block;
      padding: var(--sc-button-padding-y) var(--sc-button-padding-x);
      border-radius: calc(var(--sc-button-border-radius) - 4px);
      width: var(--sc-button-width);
    }

    &:hover,
    &:focus {
      color: var(--sc-button-text-hover-color);
      .button-inner {
        @include engraved(0.5);
      }
    }

    &:active {
      .button-inner {
        @include engraved(1);
      }
    }
  }

  &.block {
    display: block;

    a,
    button {
      width: 100%;

      .button-inner {
        width: 100%;
        text-align: center;
        justify-content: center;
      }
    }
  }

  &.icon {
    a,
    button {
      width: var(--sc-button-icon-size);
      height: var(--sc-button-icon-size);
      padding: var(--sc-button-edge-width);
      font-size: calc(var(--sc-button-icon-size) * 0.4);

      .button-inner {
        padding: 0;
        width: 100%;
        height: 100%;
        display: flex;
        align-items: center;
        justify-content: center;
      }
    }
  }
  &.icon-text {
    .button-inner {
      display: flex;
      align-items: center;
    }
  }
  &.bordered {
    a,
    button {
      position: relative;
      padding: calc(
        var(--sc-button-edge-width) + var(--sc-button-bordered-border-width)
      );
      &::before {
        content: '';
        position: absolute;
        top: calc(var(--sc-button-edge-width) - 1px);
        left: calc(var(--sc-button-edge-width) - 1px);
        bottom: calc(var(--sc-button-edge-width) - 1px);
        right: calc(var(--sc-button-edge-width) - 1px);
        border-radius: var(--sc-button-bordered-border-radius);
        @include engraved-border-gradient();
      }
      .button-inner {
        position: relative;
        background-color: var(--sc-button-bg-color);
        border-radius: calc(var(--sc-button-bordered-border-radius) - 1px);
      }
    }
    &.icon {
      a,
      button {
        --sc-button-bordered-border-radius: 50%;
        line-height: calc(
          var(--sc-button-icon-size) - (var(--sc-button-edge-width) * 2) -
            var(--sc-button-bordered-border-width) -
            var(--sc-button-bordered-border-width)
        );
      }
    }
  }
  &.circle {
    a,
    button {
      width: var(--sc-button-circle-size);
      height: var(--sc-button-circle-size);
      border-radius: 50%;
      padding: var(--sc-button-edge-width);
      &::before {
        border-radius: 50%;
      }
      .button-inner {
        padding: 0;
        width: 100%;
        height: 100%;
        display: flex;
        border-radius: 50%;
      }
    }

    &.bordered {
      a,
      button {
        padding: calc(
          var(--sc-button-edge-width) + var(--sc-button-bordered-border-width)
        );
      }
    }
  }
  &.active {
    a,
    button {
      color: var(--sc-button-text-active-color);
      .button-inner {
        @include engraved();
      }
    }
  }

  &.disabled {
    a,
    button {
      box-shadow: none;
      color: #999;
      &:hover,
      &:focus {
        .button-inner {
          box-shadow: none;
        }
      }
    }
  }
}
