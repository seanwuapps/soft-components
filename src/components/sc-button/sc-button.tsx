import {
  Component,
  Element,
  Event,
  EventEmitter,
  Host,
  Prop,
  h,
} from '@stencil/core'
/**
 * @slot - Content is placed between the named slots if provided without a slot.
 */
@Component({
  tag: 'sc-button',
  styleUrl: 'sc-button.scss',
  // shadow: true // disabled for form submission
})
export class ScButton {
  @Element() el!: HTMLElement

  /**
   * If `true`, the user cannot interact with the button.
   */
  @Prop({ reflect: true }) disabled = false

  /**
   * This attribute instructs browsers to download a URL instead of navigating to
   * it, so the user will be prompted to save it as a local file. If the attribute
   * has a value, it is used as the pre-filled file name in the Save prompt
   * (the user can still change the file name if they want).
   */
  @Prop() download: string | undefined

  /**
   * Contains a URL or a URL fragment that the hyperlink points to.
   * If this property is set, an anchor tag will be rendered.
   */
  @Prop() href: string | undefined

  /**
   * Specifies the relationship of the target object to the link object.
   * The value is a space-separated list of [link types](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types).
   */
  @Prop() rel: string | undefined

  /**
   * Make button `display: block`
   */
  @Prop({ reflect: true }) block?: boolean | undefined = false

  /**
   * Icon only button
   */
  @Prop({ reflect: true }) icon?: boolean | undefined = false

  /**
   * If prop exists, button will have an engraved-styled border
   */
  @Prop({ reflect: true }) bordered?: boolean | undefined = false

  /**
   * Specifies where to display the linked URL.
   * Only applies when an `href` is provided.
   * Special keywords: `"_blank"`, `"_self"`, `"_parent"`, `"_top"`.
   */
  @Prop() target: string | undefined

  /**
   * The type of the button.
   */
  @Prop() type: 'submit' | 'reset' | 'button' = 'button'

  /**
   * If this button has both icon and text
   */
  @Prop() iconText?: boolean | undefined = false

  /**
   * Set active state for the button
   */
  @Prop() active?: boolean = false

  /**
   * Make button flat
   */
  @Prop() flat?: boolean = false

  /**
   * Make button circle shaped
   */
  @Prop() circle?: boolean = false

  /**
   * Use mouse as the light source (ray-tracing)
   */
  @Prop() rayTracing?: boolean = false

  /**
   * Emitted when the button has focus.
   */
  @Event() focusEvent!: EventEmitter<void>

  /**
   * Emitted when the button loses focus.
   */
  @Event() blurEvent!: EventEmitter<void>

  /**
   * Emitted when the button is clicked.
   */
  @Event() clickEvent!: EventEmitter<void>

  private onFocus = () => {
    this.focusEvent.emit()
  }

  private onBlur = () => {
    this.blurEvent.emit()
  }

  private onClick = () => {
    this.clickEvent.emit()
  }

  render() {
    const {
      type,
      disabled,
      rel,
      target,
      href,
      icon,
      block,
      bordered,
      iconText,
      flat,
      circle,
    } = this
    const TagType = href === undefined ? 'button' : 'a'
    const attrs =
      TagType === 'button'
        ? { type }
        : {
            download: this.download,
            href,
            rel,
            target,
          }
    return (
      <Host
        aria-disabled={disabled ? 'true' : null}
        role="button"
        class={{
          icon,
          block,
          bordered,
          'icon-text': iconText,
          flat,
          circle,
        }}
      >
        {this.rayTracing && <sc-ray-tracer element={this.el}></sc-ray-tracer>}
        <TagType
          {...attrs}
          disabled={disabled}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onClick={this.onClick}
        >
          <span class="button-inner">
            <slot name="start"></slot>
            <slot></slot>
            <slot name="end"></slot>
          </span>
        </TagType>
      </Host>
    )
  }
}
