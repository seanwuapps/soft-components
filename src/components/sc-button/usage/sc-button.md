<sc-button onclick="alert('yo')">Button</sc-button>

<!-- Anchor -->
<sc-button href="/" target="_blank" rel="noreferrer" title="button"
  >Anchor
</sc-button>

<!-- Block level -->
<sc-button block>Block Button</sc-button>


<!-- Icon -->
<sc-button icon>
  <i class="la la-home"></i>
</sc-button>

<!-- Icon and Text -->

<sc-button iconText>
  <i class="la la-home"></i>
  Text
</sc-button>


<!-- Bordered -->
<sc-button bordered>
  <i class="la la-back"></i> Bordered button
</sc-button>


<!-- Flat -->
<sc-button flat>
  <i class="la la-back"></i> Flat button
</sc-button>