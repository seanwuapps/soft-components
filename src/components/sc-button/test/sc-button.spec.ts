import { newSpecPage } from '@stencil/core/testing'
import { ScButton } from '../sc-button'
describe('sc-button', () => {
  it('builds', () => {
    expect(new ScButton()).toBeTruthy()
  })
  it('should render default button', async () => {
    const page = await newSpecPage({
      components: [ScButton],
      html: `<sc-button>button</sc-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('should render a tag when href is applied', async () => {
    const page = await newSpecPage({
      components: [ScButton],
      html: `<sc-button href="#" target="_blank" rel="noreferrer" download>button</sc-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('should render block button when block prop is applied', async () => {
    const page = await newSpecPage({
      components: [ScButton],
      html: `<sc-button block>button</sc-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('should render icon-only button when icon prop is applied', async () => {
    const page = await newSpecPage({
      components: [ScButton],
      html: `<sc-button icon>I</sc-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('should render bordered button when bordered prop is applied', async () => {
    const page = await newSpecPage({
      components: [ScButton],
      html: `<sc-button bordered>Button</sc-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })
  it('should render a combination of different styles when props are applied', async () => {
    let page = await newSpecPage({
      components: [ScButton],
      html: `<sc-button icon bordered>Button</sc-button>`,
    })
    expect(page.root).toMatchSnapshot()

    page = await newSpecPage({
      components: [ScButton],
      html: `<sc-button block bordered>Button</sc-button>`,
    })
    expect(page.root).toMatchSnapshot()
  })
})
