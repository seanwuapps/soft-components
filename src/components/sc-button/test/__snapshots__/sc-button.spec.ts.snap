// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button should render a combination of different styles when props are applied 1`] = `
<sc-button bordered="" class="bordered icon" icon="">
  <!---->
  <button type="button">
    <span class="button-inner">
      Button
    </span>
  </button>
</sc-button>
`;

exports[`button should render a combination of different styles when props are applied 2`] = `
<sc-button block="" bordered="" class="block bordered">
  <!---->
  <button type="button">
    <span class="button-inner">
      Button
    </span>
  </button>
</sc-button>
`;

exports[`button should render a tag when href is applied 1`] = `
<sc-button download="" href="#" rel="noreferrer" target="_blank">
  <!---->
  <a download="" href="#" rel="noreferrer" target="_blank">
    <span class="button-inner">
      button
    </span>
  </a>
</sc-button>
`;

exports[`button should render block button when block prop is applied 1`] = `
<sc-button block="" class="block">
  <!---->
  <button type="button">
    <span class="button-inner">
      button
    </span>
  </button>
</sc-button>
`;

exports[`button should render bordered button when bordered prop is applied 1`] = `
<sc-button bordered="" class="bordered">
  <!---->
  <button type="button">
    <span class="button-inner">
      Button
    </span>
  </button>
</sc-button>
`;

exports[`button should render default button 1`] = `
<sc-button>
  <!---->
  <button type="button">
    <span class="button-inner">
      button
    </span>
  </button>
</sc-button>
`;

exports[`button should render icon-only button when icon prop is applied 1`] = `
<sc-button class="icon" icon="">
  <!---->
  <button type="button">
    <span class="button-inner">
      I
    </span>
  </button>
</sc-button>
`;

exports[`sc-button should render a combination of different styles when props are applied 1`] = `
<sc-button bordered="" class="bordered icon" icon="" role="button">
  <!---->
  <button type="button">
    <span class="button-inner">
      Button
    </span>
  </button>
</sc-button>
`;

exports[`sc-button should render a combination of different styles when props are applied 2`] = `
<sc-button block="" bordered="" class="block bordered" role="button">
  <!---->
  <button type="button">
    <span class="button-inner">
      Button
    </span>
  </button>
</sc-button>
`;

exports[`sc-button should render a tag when href is applied 1`] = `
<sc-button download="" href="#" rel="noreferrer" role="button" target="_blank">
  <!---->
  <a download="" href="#" rel="noreferrer" target="_blank">
    <span class="button-inner">
      button
    </span>
  </a>
</sc-button>
`;

exports[`sc-button should render block button when block prop is applied 1`] = `
<sc-button block="" class="block" role="button">
  <!---->
  <button type="button">
    <span class="button-inner">
      button
    </span>
  </button>
</sc-button>
`;

exports[`sc-button should render bordered button when bordered prop is applied 1`] = `
<sc-button bordered="" class="bordered" role="button">
  <!---->
  <button type="button">
    <span class="button-inner">
      Button
    </span>
  </button>
</sc-button>
`;

exports[`sc-button should render default button 1`] = `
<sc-button role="button">
  <!---->
  <button type="button">
    <span class="button-inner">
      button
    </span>
  </button>
</sc-button>
`;

exports[`sc-button should render icon-only button when icon prop is applied 1`] = `
<sc-button class="icon" icon="" role="button">
  <!---->
  <button type="button">
    <span class="button-inner">
      I
    </span>
  </button>
</sc-button>
`;
